### PR TITLE
Handle inter-expansion top-level args properly

### LINKED ISSUE #113: struct annotation doesn't work on a flatten struct
example at https://github.com/rust-lang-nursery/cli-wg/issues/37#issuecomment-392527822

### LINKED ISSUE #151: Groups within subcommands not working as expected
I am trying to get a program with subcommands, one of which requires at least one flag. So, for example, I would want the following behavior:
* `cargo run -- a` fails
* `cargo run -- a -c` succeeds
* `cargo run -- a -d` succeeds
* `cargo run -- a -cd` succeeds

I can get it working using a `clap::ArgGroup` without subcommands, but not with subcommands.
For example, given the following
```rust
use structopt::clap::ArgGroup;
use structopt::StructOpt;

fn arg_group() -> ArgGroup<'static> {
    ArgGroup::with_name("group").required(true).multiple(true)
}

#[derive(StructOpt, Debug)]
#[structopt(raw(group = "arg_group()"))]
struct Sub {
    #[structopt(short = "c", group = "group")]
    c: bool,
    #[structopt(short = "d", group = "group")]
    d: bool,
}

#[derive(StructOpt, Debug)]
enum Opt {
    #[structopt(name = "a", raw(group = "arg_group()"))]
    A (Sub),
    #[structopt(name = "b")]
    B {},
}
```
this works (though not as a subcommand)
```rust
fn main() {
    let opt = Sub::from_args();
    println!("{:?}", opt);
}
```
but this does not
```rust
fn main() {
    let opt = Opt::from_args();
    println!("{:?}", opt);
}
```

Instead, I get the following behavior:
* `cargo run -- a` fails
* `cargo run -- a -c` succeeds
* `cargo run -- a -d` succeeds
* `cargo run -- a -cd` fails

So I'm not sure if I'm doing something wrong or if perhaps its a bug in structopt.

I'm using structopt 0.2.13 and rust nightly (6acbb5b65 2018-11-25).

### LINKED ISSUE #289: 'infer subcommand' doesn't work with nested subcommands
When `InferSubcommands` is enabled, subcommand inference only works for the top level of subcommands- to call nested subcommands the entire command must be written out.

```rust
use structopt::{StructOpt, clap::AppSettings};

#[derive(StructOpt)]
#[structopt(setting = AppSettings::InferSubcommands)]
enum Args {
    SomeCommand(SubSubCommand),
    AnotherCommand,
}

#[derive(StructOpt)]
#[structopt(setting = AppSettings::InferSubcommands)]
enum SubSubCommand {
    TestCommand,
}

fn main() {
    let args = Args::from_args();
}
```

- `cargo run some-command test-command` *works*
- `cargo run some test-command` *works*
- `cargo run some-command test` *doesn't work*
- `cargo run some test` *doesn't work*

i've tried both the above syntax, and explicitly marking with the `#[structopt(subcommand]` tag.

### PR DESCRIPTION
**Pull request by CreepySkeleton**
Fixes #151 
Fixes #289 
Fixes #113

### CODE REVIEW

**Comment by CreepySkeleton** [comment]
@TeXitoi What's breaking with it?

**Comment by TeXitoi** [comment]
just doing a cargo update will change the behavior of the program as the top level annotations of parent types will be executed.

**Comment by CreepySkeleton** [comment]
> top level annotations of parent types will be executed.

Well, I *really* doubt someone out there would put an attribute and *expect it not to apply* :)

**Comment by TeXitoi** [comment]
That's why i'm OK with that, but it can happen as someone might tests some things and do not remove top level attribute but is happy with the result.

**Comment by TeXitoi** [comment]
It can be considered as a bug.

**Comment by CreepySkeleton** [comment]
> do not remove top level attribute but is happy with the result.

You don't exactly expect such careless programming yield good results :)

**Comment by TeXitoi** [comment]
Welcome to the real world! ;-)

File: tests/issues.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    assert!(Cli::clap().get_matches_from_safe(&["test"]).is_err());
+    assert!(Cli::clap()
+        .get_matches_from_safe(&["test", "--foo"])
+        .is_ok());
+    assert!(Cli::clap()
+        .get_matches_from_safe(&["test", "--bar"])
+        .is_ok());
+    assert!(Cli::clap()
+        .get_matches_from_safe(&["test", "--zebra"])
+        .is_err());
```
[the remark this answers is not in the document]
Comment (CreepySkeleton) [reply tests/issues.rs]: @Veetaha , just so you know, github doesn't send notifications for review that's left by non maintainer, or at least I didn't receive it. Honestly, I just hit a wrong button and stumbled on this :) Please use explicit mentioning, like @CreepySkeleton , if you want to be noticed.

**Pull request merged**